src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -345,6 +345,112 @@ void QgsPostgresProvider::setTransaction( QgsTransaction *transaction )
   mTransaction = static_cast<QgsPostgresTransaction *>( transaction );
 }
 
+struct Ewkt
+{
+  int srid = -1;
+  QString wkt;
+};
+
+QgsReferencedGeometry QgsPostgresProvider::fromEwkt( const QString &ewkt, QgsPostgresConn *conn )
+{
+  thread_local const QRegularExpression regularExpressionSRID( "^SRID=(\\d+);" );
+
+  QRegularExpressionMatch regularExpressionMatch = regularExpressionSRID.match( ewkt );
+  if ( !regularExpressionMatch.hasMatch() )
+    return QgsReferencedGeometry();
+
+  Ewkt ewktInfo;
+  ewktInfo.wkt = ewkt.mid( regularExpressionMatch.captured( 0 ).size() );
+  ewktInfo.srid = regularExpressionMatch.captured( 1 ).toInt();
+
+  if ( ewktInfo.srid < 0 )
+    return QgsReferencedGeometry();
+
+  QgsGeometry geom = QgsGeometry::fromWkt( ewktInfo.wkt );
+  return QgsReferencedGeometry( geom, sridToCrs( ewktInfo.srid, conn ) );
+}
+
+QString QgsPostgresProvider::toEwkt( const QgsReferencedGeometry &geom, QgsPostgresConn *conn )
+{
+  return QStringLiteral( "SRID=%1;%2" ).arg( QString::number( crsToSrid( geom.crs(), conn ) ), geom.asWkt() );
+}
+
+QString QgsPostgresProvider::geomAttrToString( const QVariant &attr, QgsPostgresConn *conn )
+{
+  if ( attr.type() == QVariant::String )
+    return attr.toString();
+  else
+    return toEwkt( attr.value<QgsReferencedGeometry>(), conn );
+}
+
+static QMutex sMutex;
+static QMap<int, QgsCoordinateReferenceSystem> sCrsCache;
+
+int QgsPostgresProvider::crsToSrid( const QgsCoordinateReferenceSystem &crs, QgsPostgresConn *conn )
+{
+  QMutexLocker locker( &sMutex );
+  int srid = sCrsCache.key( crs );
+
+  if ( srid > -1 )
+    return srid;
+  else
+  {
+    if ( conn )
+    {
+      QStringList authParts = crs.authid().split( ':' );
+      if ( authParts.size() != 2 )
+        return -1;
+      const QString authName = authParts.first();
+      const QString authId = authParts.last();
+      QgsPostgresResult result( conn->PQexec( QStringLiteral( "SELECT srid FROM spatial_ref_sys WHERE auth_name='%1' AND auth_srid=%2" ).arg( authName, authId ) ) );
```

Comment:
@m-kuhn , is this spatial_ref_sys table normally large? i.e., is it the whole of spatial references ala spatialite or only what the database uses?

Asking as I'm wondering whether we should just fetch the whole thing once when first needed.

Reply:
I guess it's a very common scenatio that you only need one, so I'm not sure it's worth this upfront effort by default. In favor of keeping provider init times low.

Reply:
I agree with @m-kuhn here -- I think we should avoid loading the whole table